### PR TITLE
Support for sub flows in defining content enricher

### PR DESCRIPTION
**Pull request by chiknrice**
The current enricher allows for specifying a payload, request and response channel where one can request additional data for enriching content or header. This change would allow one to specify an inline call to a handler instead of a request/response channel to grab the additional data.

### CODE REVIEW

File: spring-integration-core/src/main/java/org/springframework/integration/dsl/IntegrationFlowDefinition.java
Comment (artembilan) [inline spring-integration-core/src/main/java/org/springframework/integration/dsl/IntegrationFlowDefinition.java]: To be honest I don't see reason in this restriction.
There is no guarantee that flow ends with just a reply producer. It may sends to the channel and a flow behind it may return the result to the `replyChannel`.

Yes, that is the point to wait for the reply from the downstream flow for the `enricher`, but similar concern with unpredictable sub-flow behavior we have even with an existing XML definition.
So, my point to relax this assertion and really just configure `requestChannel` as an input to the that sub-flow and 
```
this.componentsToRegister.add(flowBuilder.get());
```
in the `EnricherSpec`.

That way this `IntegrationFlowDefinition` won't be affected at all.

The problem with "no reply from sub-flow" remains to the end-user. 
No choice!

File: spring-integration-core/src/main/java/org/springframework/integration/dsl/EnricherSpec.java
Comment (artembilan) [inline spring-integration-core/src/main/java/org/springframework/integration/dsl/EnricherSpec.java]: We don't need to call `_this()` here

**Review by artembilan — CHANGES_REQUESTED** [review]
@chiknrice ,
Thank you very much for the contribution!

See comments bellow.
Only several simple concerns.
No reason to comment if you agree with me.
Just push the fix and that's all.
I'll merge it!

One more comment: please, add your name to the `@author` of affected classes.

File: spring-integration-core/src/main/java/org/springframework/integration/dsl/IntegrationFlowDefinition.java
Comment (chiknrice) [inline spring-integration-core/src/main/java/org/springframework/integration/dsl/IntegrationFlowDefinition.java]: Agreed, I didn't thought of that :+1: also added a test of a terminating sub-flow

File: spring-integration-core/src/test/java/org/springframework/integration/dsl/transformers/TransformerTests.java
Comment (artembilan) [inline spring-integration-core/src/test/java/org/springframework/integration/dsl/transformers/TransformerTests.java]: Not sure why do you worry about `async` behavior in the sub-flow, since `ContentEnricher` uses blocking `sendAndReceiveMessage()`. So, it doesn't matter how the reply is going to appear. But yeah, it won't hurt to have that test as well. 
M-m-m. I believe that `@Autowired` is redundant on the `@Bean` method. Any its arguments are really considered for autowiring.

File: spring-integration-core/src/test/java/org/springframework/integration/dsl/transformers/TransformerTests.java
Comment (artembilan) [inline spring-integration-core/src/test/java/org/springframework/integration/dsl/transformers/TransformerTests.java]: Let's remove this bean altogether!
The `GatewayProxyFactoryBean` uses its instance by default anyway:
```
private volatile AsyncTaskExecutor asyncExecutor = new SimpleAsyncTaskExecutor();
```
This test isn't about Messaging Gateway, so, would be better to make it as simple as possible.

File: spring-integration-core/src/test/java/org/springframework/integration/dsl/transformers/TransformerTests.java
Comment (artembilan) [inline spring-integration-core/src/test/java/org/springframework/integration/dsl/transformers/TransformerTests.java]: I think you are aware of the `@MessagingGateway`  and `@IntegrationComponentScan`.
Maybe we really don't need this bean definition?

File: spring-integration-core/src/test/java/org/springframework/integration/dsl/transformers/TransformerTests.java
Comment (artembilan) [inline spring-integration-core/src/test/java/org/springframework/integration/dsl/transformers/TransformerTests.java]: I think this `SomeService` can be marked with the `@Service` as well.
Yes, `@ComponentScan` must be there on the `@Configuration`.

That feels simpler.
Just IMO any way.
We can live with the current state of the test.
Looks like it demonstrates the premise of the feature and that is enough.

**Review by artembilan — CHANGES_REQUESTED** [review]
@chiknrice ,

thank you for the update!

Please, meet the next review round.
I don't see other issues any more and after your next push this might be merged.

File: spring-integration-core/src/test/java/org/springframework/integration/dsl/transformers/TransformerTests.java
Comment (chiknrice) [inline spring-integration-core/src/test/java/org/springframework/integration/dsl/transformers/TransformerTests.java]: I initially had the requirement of enriching a message with information coming from a web resource and was thinking it was a good idea to make it an async call.  I realise now it doesn't makes sense since the whole enrich call is blocking - maybe i'm approaching the problem incorrectly.  Anyway, will still submit this change as it could be helpful to some.  I've applied the necessary changes to the test related to this.

File: spring-integration-core/src/test/java/org/springframework/integration/dsl/transformers/TransformerTests.java
Comment (chiknrice) [inline spring-integration-core/src/test/java/org/springframework/integration/dsl/transformers/TransformerTests.java]: I removed the beans relating to flow with async handler and I think the test was designed to not do component scanning.

File: spring-integration-core/src/test/java/org/springframework/integration/dsl/transformers/TransformerTests.java
Comment (chiknrice) [inline spring-integration-core/src/test/java/org/springframework/integration/dsl/transformers/TransformerTests.java]: Done.

**Review by artembilan — APPROVED** [review]
LGTM
Merging...

**Comment by artembilan** [comment]
Merged as https://github.com/spring-projects/spring-integration/commit/6616242831a9634fd8f9e260ccc448406e98522b.

Thank you very much for the contribution.
Looking forward for more!

**Pull request closed without merging**
